apps/labs/posts/numpy-string-ufuncs.md

```diff
@@ -0,0 +1,546 @@
+integer). Thus, all integer inputs should be promoted to int64 before searching for
+a loop. Our promoter will then look like this:
+
+```cpp
+int
+string_find_rfind_promoter(PyUFuncObject *ufunc,
+                           PyArray_DTypeMeta *op_dtypes[],
+                           PyArray_DTypeMeta *signature[],
+                           PyArray_DTypeMeta *new_op_dtypes[])
+{
+    // Do not promote the first two arguments (haystack and needle)
+    Py_INCREF(op_dtypes[0]);
+    new_op_dtypes[0] = op_dtypes[0];
+    Py_INCREF(op_dtypes[1]);
+    new_op_dtypes[1] = op_dtypes[1];
+
+    // Promote any other type to int64 for the next two arguments (indices)
+    new_op_dtypes[2] = PyArray_DTypeFromTypeNum(NPY_INT64);
+    new_op_dtypes[3] = PyArray_DTypeFromTypeNum(NPY_INT64);
+
+    // Use default integer as the output type
+    new_op_dtypes[4] = PyArray_DTypeFromTypeNum(NPY_DEFAULT_INT);
+    return 0;
+}
+```
+
+Our promoter specifies the following:
+1. For the two string arguments, we do not allow promotion. We take the input dtype
+   as is.
+2. For the two integer inputs, we allow other dtypes and promote all of them to int64.
+3. The output dtype will be the default integer dtype.
+
+Doing all of this allows us to do the following, having only written one loop:
+
+```cpp
+>>> buf = np.array(["abcd", "cdef"])
+>>> sub = "c"
+>>> start = np.array([3, 0], dtype=np.int8)
+>>> end = np.array([4, 4], dtype=np.int8)
+>>> np.strings.find(buf, sub, start, end)
+array([-1,  0])
+>>> start = np.array([3, 0], dtype=np.int16)
+>>> end = np.array([4, 4], dtype=np.int16)
+>>> np.strings.find(buf, sub, start, end)
+array([-1,  0])
+```
+
+## Last stop: `replace`
+
+The last ufunc we're going to talk about is `replace`. First, let's remember that
+`replace` accepts four arguments: the buffer to search, the substring to search for
+and replace, the replacement string, and a final integer argument that specifies
+the maximum number of replacements to do. Putting everything we've covered so far
+together gives us the following tasks:
+
+1. Write a loop for the string dtypes and int64.
+2. Add a promoter that promotes other integer dtypes to int64 for the last argument.
+3. Write a `resolve_descriptors` function that specifies the output dtype size.
+
+But, wait! Those that paid attention to the section about `add` will remember that
+the `resolve_descriptors` function does not get access to the data itself; it only
+gets access to the array descriptors that contain meta information about the array.
+By just looking at the input dtype sizes though, we do not know how many replacements
+we'll have to do, and how long the final string will be. In order to circumvent
+that, we have to create a Python wrapper around the ufunc written in C++. But,
+before we explain how the wrapper works, let's talk about another feature of ufuncs:
+the `out` optional keyword argument.
+
+If the optional `out` keyword argument is `None`, which is the default, a new array
```

Comment:
Maybe worth describing what `out` does earlier in the blog where you go over `add` above, since this is a more fundamental concept that readers who are familiar with numpy might know about, but probably don't know how it's handled in a ufunc implementation.

Reply:
Moved to the `add` section. Maybe worth checking that it doesn't look out of place there.